src/libostree/ostree-bootloader-zipl.c

```diff
@@ -80,6 +83,203 @@ _ostree_bootloader_zipl_write_config (OstreeBootloader  *bootloader,
+  while ((line = g_data_input_stream_read_line (data, NULL, NULL, NULL))) {
+      gchar **values = g_strsplit (line, " ", 2);
+      if (g_str_has_prefix (line, "linux ")) {
+          *vmlinuz = g_string_new ("/boot");
+          g_string_append (*vmlinuz, values[1]);
+          sd_journal_print(LOG_INFO, "s390x SE: kernel: %s", (*vmlinuz)->str);
+        } else if (g_str_has_prefix (line, "initrd ")) {
+          *initram = g_string_new ("/boot");
+          g_string_append (*initram, values[1]);
+          sd_journal_print(LOG_INFO, "s390x SE: initrd: %s", (*initram)->str);
+        } else if (g_str_has_prefix (line, "options ")) {
+          *options = g_string_new (values[1]);
+          sd_journal_print(LOG_INFO, "s390x SE: kargs: %s", (*options)->str);
+        }
+      g_strfreev (values);
+      free (line);
+    }
+
+  return TRUE;
+}
+
+static gboolean
+_ostree_secure_execution_generate_sdboot (GString *vmlinuz,
+                                          GString *initram,
+                                          GString *options,
+                                          GArray *keys,
+                                          GError **error)
+{
+  if (!vmlinuz || !initram || !options || !keys)
+    {
+      sd_journal_print(LOG_ERR, "s390x SE: vmlinuz, initram, options, keys are missing");
+      return FALSE;
+    }
+
+  g_autoptr(GString) cmd   = NULL;
+  gboolean rc              = FALSE;
+  g_autofree char *cmdline = g_strdup ("/tmp/sd_boot.parmfile.XXXXXX");
+  gint fd                  = g_mkstemp (cmdline);
+  if (glnx_loop_write (fd, options->str, options->len) < 0)
+    {
+      sd_journal_print(LOG_ERR, "s390x SE: creating %s: %s", cmdline, g_strerror(errno));
+      goto out;
+    }
+
+  cmd = g_string_new ("genprotimg");
+  g_string_append_printf (cmd, " -i %s", vmlinuz->str);
+  g_string_append_printf (cmd, " -r %s", initram->str);
+  g_string_append_printf (cmd, " -p %s", cmdline);
+  for (guint i = 0; i < keys->len; ++i) {
+      GString *key = g_array_index (keys, GString*, i);
+      g_string_append_printf (cmd, " -k %s", key->str);
+    }
+  g_string_append (cmd, " --no-verify");
+  g_string_append (cmd, " -o " SECURE_EXECUTION_BOOT_IMAGE);
+
+  gint status = 0;
+  if (!g_spawn_command_line_sync (cmd->str, NULL, NULL, &status, error))
+    {
+      sd_journal_print(LOG_INFO, "s390x SE: executing `genprotimg`: %s", (*error)->message);
+      goto out;
+    }
+
+  if (!g_spawn_check_exit_status (status, error))
+    {
+      sd_journal_print(LOG_INFO, "s390x SE: `genprotimg` failed: %s", (*error)->message);
+      goto out;
+    }
+  rc = TRUE;
+  sd_journal_print(LOG_INFO, "s390x SE: `sd-boot` generated");
+
+out:
+  unlink(cmdline);
+  close(fd);
+
+  return rc;
+}
+
+static gboolean
+_ostree_secure_execution_call_zipl (GError **error)
+{
+  int status;
+  const char *cmd = "zipl -V -t /boot -i " SECURE_EXECUTION_BOOT_IMAGE;
+  if (!g_spawn_command_line_sync (cmd, NULL, NULL, &status, error))
```

Comment:
In ostree we usually avoid involving the shell for executing subprocesses.  If you do, it's important to use e.g. `g_shell_quote()` to avoid quoting problems.

But here I'd use `g_spawn_sync()`; there are other examples in the codebase.